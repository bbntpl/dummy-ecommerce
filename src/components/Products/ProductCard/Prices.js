import {
	Price,
	PrevPrice,
} from './product-card-styling';
import DecimalPrecision from '../../../js/decimalPrecision';
import { discountedPrice } from '../../../js/reusableFuncs';
export default function ProductPrices(props) {
	const { discount, price } = props;
	const roundedPrice = DecimalPrecision.round(discountedPrice(price, discount), 2);
	return (
		<span>
			<Price>
				${discount ? roundedPrice : price}
			</Price>
			{
				discount &&
				<PrevPrice>
					${price}
				</PrevPrice>
			}
		</span>
	)
}