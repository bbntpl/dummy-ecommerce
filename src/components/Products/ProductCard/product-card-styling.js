import styled from 'styled-components';
import { Segment, Header } from 'semantic-ui-react';
import { LazyLoadImage } from 'react-lazy-load-image-component';

const productCardStyles = `
	display: flex!important;
	flex-direction: column!important;
	width: 100%;
	padding: 10px 0;
`;

const ThumbnailStyles = `
	height: auto;
	max-height: 300px;
	width: fit-content;
	max-width: 100%;
`

export const StyledProductCard = styled(Segment)`
	${productCardStyles}
	justify-content: space-between;
	height: 100%;
	& h3 {
		font-size: 14px;
	}
	& > div:first-child {
		display: flex;
		justify-content: center;
	}
`

export const Details = styled(StyledProductCard)`
	background-color: hsla(0,0%,40%,.08)!important;
	font-weight: 700!important;
	gap: 5px;
	width: 100%;
	align-items: center!important;
	height: max-content;
	& > a {
		text-align: center;
	}
	& > a:hover {
		text-decoration: underline;
		& > div {
			color: #4183C4!important;
		}
	}
	& > .product-card__discount {
		padding: 4px 8px;
		border: 2px solid #141348;
		border-radius: 12px;
		font-size: 1rem;
		width: fit-content;
		font-weight: 500;
		background: linear-gradient(45deg, gold, GreenYellow);
	}
`;

export const StyledThumbnail = styled(LazyLoadImage)`
	${ThumbnailStyles}
`
export const Title = styled(Header)`
	display: inline!important;
	font-size: clamp(1rem 3vw 3rem)!important;
	color: #4D0000;
	font-weight: 700!important;
`

export const Price = styled.p`
	display: inline-block;
	font-size: 20px;
	color: #3F3F3F;
`

export const PrevPrice = styled.p`
	display: inline-block;
	padding-left: 13px;
	font-weight: 400;
	color: #919191;
	text-decoration: line-through;
`