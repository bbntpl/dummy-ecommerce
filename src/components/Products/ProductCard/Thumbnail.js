import { memo } from 'react';
import ImagePlaceholderSrc from '../../../assets/images/image-placeholder.png';
import { Link } from 'react-router-dom';
import { Image } from 'semantic-ui-react';
import { StyledThumbnail } from './product-card-styling';

function Thumbnail({ src, category, id }) {
	return (
		<div>
			<Link to={`product/${id}`}>
				<StyledThumbnail
					alt={`${category}${id}`}
					src={src}
					effect='opacity'
					height={'auto'}
					placeholder={
						<Image
							src={ImagePlaceholderSrc}
							className='ui medium image'
						/>
					}
				/>
			</Link>
		</div>
	);
}

const MemoizedThumbnail = memo(Thumbnail);
export default MemoizedThumbnail;
