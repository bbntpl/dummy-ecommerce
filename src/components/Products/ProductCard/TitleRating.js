import { Link } from 'react-router-dom';
import { Title } from './product-card-styling';

export default function TitleRating(props) {
	const { title, rating, id } = props;
	return (
		<Link to={`product/${id}`}>
			<Title>
				{`${title} - `}
			</Title>
			<span> â˜… {rating}</span>
		</Link>
	)
}