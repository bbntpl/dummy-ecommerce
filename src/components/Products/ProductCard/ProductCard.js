import { forwardRef } from 'react';
import { Button, Icon } from 'semantic-ui-react';
import { StyledProductCard, Details } from './product-card-styling';
import TitleRating from './TitleRating';
import Prices from './Prices';
import Thumbnail from './Thumbnail';
import PlaceholderImage from '../../../assets/images/image-placeholder.png';

const ProductCard = forwardRef((props, ref) => {
	const { product, addItemToCart, getItemQty } = props;
	const {
		id,
		title,
		thumbnail,
		price,
		rating,
		stock,
		category,
		discountPercentage,
	} = product;
	const isStockMoreThanQty = stock > getItemQty(id);

	// disallow to add item if quantity is more than the stock
	const addItemToCartIfAvailable = (id) => {
		if (!isStockMoreThanQty) return;
		addItemToCart({ targetId: id });
	}
	return (
		<div ref={ref} style={{ height: '100%' }}>
			<StyledProductCard>
				<Thumbnail 
				src={thumbnail} 
				category={category} 
				id={id} 
				placeholder={PlaceholderImage}
				/>
				<Details>
					<TitleRating
						title={title}
						id={id}
						rating={rating}
					/>
					<h3 className='product-card__discount'>{`SAVE ${discountPercentage}%`}</h3>
					<Prices price={price} discount={discountPercentage} />
					<Button
						onClick={() => addItemToCartIfAvailable(id)}
						size='large'
						disabled={!isStockMoreThanQty}
					>
						<Icon name='shopping cart' />
						{isStockMoreThanQty ? 'Add to cart' : 'Not enough stock'}
					</Button>
				</Details>
			</StyledProductCard>
		</div>
	)
})

export default ProductCard;