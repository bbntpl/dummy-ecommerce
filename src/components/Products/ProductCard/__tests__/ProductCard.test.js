import { render, screen, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { BrowserRouter as Router } from 'react-router-dom';
import ProductCard from '../index';

const product = {
	discountPercentage: 40,
	id: 1,
	category: 'phone',
	price: 99.99,
	rating: 4.69,
	stock: 4,
	thumbnail: 'https://dummyjson.com/image/i/products/1/thumbnail.jpg',
	title: 'iPhone 9',
	quantity: 0,
}

const addItemToCart = () => {
	product.quantity += 1;
}

const getItemQty = () => {
	return product.quantity;
}

it('should render elements with written output', () => {
	const { container } = render(
		<Router>
			<ProductCard
				product={product}
				addItemToCart={addItemToCart}
				getItemQty={getItemQty}
			/>
		</Router>
	);

	expect(container).toMatchSnapshot();
});

it('disable button and replace text with not enough stock', () => {
	const RenderProductCard = () => (
		<Router>
			<ProductCard
				product={product}
				addItemToCart={addItemToCart}
				getItemQty={getItemQty}
			/>
		</Router>
	)
	const { rerender } = render(<RenderProductCard />);
	const button = screen.getByRole('button');

	act(() => {
		for (let i = 0; i < product.stock; i++) {
			userEvent.click(button);
		}
	});

	rerender(<RenderProductCard />)
	expect(button).toHaveTextContent('Not enough stock');
	expect(button).toBeDisabled();
});