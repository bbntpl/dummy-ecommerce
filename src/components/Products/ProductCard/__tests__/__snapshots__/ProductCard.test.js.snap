// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render elements with written output 1`] = `
<div>
  <div
    style="height: 100%;"
  >
    <div
      class="ui segment sc-bczRLJ hhAjhe"
    >
      <div>
        <a
          href="/product/1"
        >
          <span
            class=" lazy-load-image-background opacity"
            style="color: transparent; display: inline-block; height: auto;"
          >
            <img
              alt="phone1"
              class="sc-dkzDqf cKdirU"
              height="auto"
              src="https://dummyjson.com/image/i/products/1/thumbnail.jpg"
            />
          </span>
        </a>
      </div>
      <div
        class="ui segment sc-bczRLJ sc-gsnTZi hhAjhe eOIndq"
      >
        <a
          href="/product/1"
        >
          <div
            class="ui header sc-hKMtZM ezvbUU"
          >
            iPhone 9 - 
          </div>
          <span>
             â˜… 
            4.69
          </span>
        </a>
        <h3
          class="product-card__discount"
        >
          SAVE 40%
        </h3>
        <span>
          <p
            class="sc-eCYdqJ idwgtc"
          >
            $
            59.99
          </p>
          <p
            class="sc-jSMfEi jzdFeH"
          >
            $
            99.99
          </p>
        </span>
        <button
          class="ui large button"
        >
          <i
            aria-hidden="true"
            class="shopping cart icon"
          />
          Add to cart
        </button>
      </div>
    </div>
  </div>
</div>
`;
